core/src/Cabana_Parallel.hpp

```diff
@@ -1059,6 +1061,217 @@ for_each_neighbor( const IndexType i, const TeamMemberType team,
+  \param functor The functor to execute in parallel
+  \param list The linked cell list over which to execute neighbor operations.
+  \param cell_stencil Stencil of cell neighbors to iterate over.
+  \param FirstNeighborsTag Tag indicating operations over particle first
+  neighbors.
+  \param SerialOpTag Tag indicating a serial loop strategy over neighbors.
+  \param str Optional name for the functor. Will be forwarded if non-empty to
+  the Kokkos::parallel_for called by this code and can be used for
+  identification and profiling purposes.
+
+  A "functor" is a class containing the function to execute in parallel, data
+  needed for that execution, and an optional \c execution_space typedef.  Here
+  is an example functor for neighbor parallel_for:
+
+  \code
+  class FunctorType {
+  public:
+  typedef  ...  execution_space ;
+  void operator() ( const int particle_index, const int neighbor_index ) const ;
+  };
+  \endcode
+
+  In the above example, \c Index is a Cabana index to a given AoSoA element
+  for a particle and its neighbor. Its <tt>operator()</tt> method defines the
+  operation to parallelize, over the range of indices
+  <tt>idx=[begin,end]</tt>.  This compares to a single iteration \c idx of a
+  \c for loop.
+*/
+template <class FunctorType, class LinkedCellType, class StencilType,
+          class... ExecParameters>
+inline void linkedcell_parallel_for(
+    const Kokkos::RangePolicy<ExecParameters...>& exec_policy,
+    const FunctorType& functor, const LinkedCellType& list,
+    const StencilType& cell_stencil, const FirstNeighborsTag, const SerialOpTag,
+    const std::string& str = "" )
+{
+    using work_tag = typename Kokkos::RangePolicy<ExecParameters...>::work_tag;
+
+    using execution_space =
+        typename Kokkos::RangePolicy<ExecParameters...>::execution_space;
+
+    using index_type =
+        typename Kokkos::RangePolicy<ExecParameters...>::index_type;
+
+    using memory_space = typename LinkedCellType::device_type::memory_space;
+
+    NeighborDiscriminator<FullNeighborTag> discriminator;
+
+    auto begin = exec_policy.begin();
+    auto end = exec_policy.end();
+    using linear_policy_type = Kokkos::RangePolicy<execution_space, void, void>;
+    linear_policy_type linear_exec_policy( begin, end );
+
+    static_assert( is_accessible_from<memory_space, execution_space>{}, "" );
+
+    auto cell = list.cells( begin, end );
+    auto neigh_func = KOKKOS_LAMBDA( const index_type i )
+    {
+        int imin, imax, jmin, jmax, kmin, kmax;
+        cell_stencil.getCells( cell( i ), imin, imax, jmin, jmax, kmin, kmax );
+
+        // Loop over the cell stencil.
+        for ( int gi = imin; gi < imax; ++gi )
+            for ( int gj = jmin; gj < jmax; ++gj )
+                for ( int gk = kmin; gk < kmax; ++gk )
+                {
+                    // Check the particles in this bin to see if they are
+                    // neighbors.
+                    auto offset = list.binOffset( gi, gj, gk );
```

Comment:
I think you could avoid all this if you use at the neighbor list interface directly. You can still call the discriminator but I think we would honestly want to leave that up to user still because this could be used for MD, etc. Based on that I'm not sure you need this overload at all.

Reply:
Planning to test this and compare performance (doing all this extra indexing for each thread may be a lot)